### PR TITLE
functionMap: fixed return type of getopt()

### PR DESCRIPTION
**Pull request by dktapps**
this can return FALSE if $argv is not an array for some reason (maybe register_argv_argv was disabled, or maybe it was overwritten by user code).

This isn't currently reported on 7.x, but is reported on 8.0 (due to stubs differences).

### CODE REVIEW

**Comment by ondrejmirtes** [comment]
Dynamic return type extension would be much better fit.

**Comment by dktapps** [comment]
I'm not sure how to implement a dynamic return type extension for this. The result depends on global variables and INI values.

**Comment by ondrejmirtes** [comment]
We already have `%cliArgumentsVariablesRegistered%` that influences some things so we could use that.

**Comment by dktapps** [comment]
`config.neon` is a bundle of surprises. I had no idea that setting was lurking in there.

**Comment by dktapps** [comment]
The behaviour of this function is way more complex than I thought. I'm not sure that this is the only case where it returns `false`. Creating a dynamic extension for this is going to take some time.

**Comment by ondrejmirtes** [comment]
We could get by just by writing an extension that calls `TypeUtils::toBenevolentUnion()` on the already existing return type. That way PHPStan won't complain about comparison against `false`.

**Comment by dktapps** [comment]
Will that also work on PHP 8.0? The native union already specifies `array|false`.

**Comment by ondrejmirtes** [comment]
Should. Feel free to try.

**Comment by dktapps** [comment]
It seems like the cases where `false` can be returned are:
- when `register_argc_argv` is turned `off`
- when `$argv` is set to `NULL` in the global scope

When `register_argc_argv` is disabled, `$argv` is null anyway. The problems are:
- PHPStan doesn't know that `$argv` has a type of `array` outside of `DefinedVariableRule`. I'm not sure where I should add the code for this (we essentially need to define two additional global variables for PHPStan's context when `register_argc_argv` is enabled).
- It's impossible to keep track of the state of globals (they might be nulled any time by user code).

So, I don't think there's any pathway to solving this problem short of either checking if `$argv` is an array in user code, or making this union type benevolent so that it won't complain (because it's at best going to be `array|false`, and at worst `false`).

**Comment by dktapps** [comment]
I think this is good to merge. I might try and contribute a more complete extension in the future, but this should be sufficient to solve the `false` annoyances.

File: src/Type/Php/GetoptFunctionDynamicReturnTypeExtension.php [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+	public function isFunctionSupported(FunctionReflection $functionReflection): bool
+	{
+		return $functionReflection->getName() === 'getopt';
+	}
+
+	public function getTypeFromFunctionCall(FunctionReflection $functionReflection, FuncCall $functionCall, Scope $scope): Type
+	{
+		return TypeUtils::toBenevolentUnion(TypeCombinator::union(
+			ParametersAcceptorSelector::selectSingle($functionReflection->getVariants())->getReturnType(),
+			new ConstantBooleanType(false)
```
Comment (ondrejmirtes) [inline src/Type/Php/GetoptFunctionDynamicReturnTypeExtension.php:27]: You don't have to `union` with `false` here as it's already there in the functionMap.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/Type/Php/GetoptFunctionDynamicReturnTypeExtension.php (hunk 1 above)
Comment (dktapps) [reply src/Type/Php/GetoptFunctionDynamicReturnTypeExtension.php:27]: I was under the impression that the return type extension overrode the function map type entirely.

**Comment by ondrejmirtes** [comment]
No, it's in the function reflection variant's return type.

**Comment by ondrejmirtes** [comment]
Thank you!

**Comment by ondrejmirtes** [comment]
FYI you actually broke analysis on your own code here :) https://github.com/phpstan/phpstan/runs/1776481252?check_suite_focus=true#step:6:109

**Comment by dktapps** [comment]
Actually that's how I found this bug in the first place. I'm working on 8.0 migration and this issue showed up on 8.0 but not 7.4.

**Comment by ondrejmirtes** [comment]
Sorry, had to revert it so I can make a release.

**Comment by dktapps** [comment]
Why?

**Comment by ondrejmirtes** [comment]
Because it doesn't work for analysing PMMP codebase therefore I figured it'd cause more false positives than being useful.

**Comment by dktapps** [comment]
Well, it causes an inconsistency between 7.4 and 8.0, that's why I pulled this to begin with (I want to minimize the amount of specialization I have to do for 8.0). All of those errors appear on PM in 8.0 in any case. https://github.com/pmmp/PocketMine-MP/runs/1810957332?check_suite_focus=true#step:9:278

**Comment by aadmathijssen** [comment]
Hi, I ran into this issue today: according to the PHP documentation `getopt` might return `false`, so I need to check this scenario. However, phpstan fails with the following message:

```
Strict comparison using === between array<string, array<int, mixed>|string|false> and false will always evaluate to false.
```

The strange thing is that it only does this on PHP 7.4, but not on PHP 8.0.

For a minimal example, see https://phpstan.org/r/d472f1e8-6afa-4e7a-a9aa-e7a0258a576c.

So merging this PR (or a variant of it) would be appreciated.

Thanks!

**Comment by dktapps** [comment]
@aadmathijssen this is exactly why I made the PR, to resolve the inconsistency between 7.4 and 8.0 :<

**Pull request merged**